### PR TITLE
Do not allow custom analyzers to have the same names as built-in analyzers

### PR DESCRIPTION
**Pull request by makeyang**
throw execption when custom analyzers have the same names as built-in analyzers according to issue:
https://github.com/elastic/elasticsearch/issues/22263

### CODE REVIEW

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 1]
```diff
@@ -271,6 +271,9 @@ public boolean requiresAnalysisSettings() {
             Settings currentSettings = entry.getValue();
             String typeName = currentSettings.get("type");
             if (analyzer) {
+                if(defaultInstance.containsKey(name)) {
+                    throw new IllegalArgumentException("do not allow custom analyzers to have the same names [" + name + "] as built-in analyzers");
+                }
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: We need to only enforce this on new indices, can you add a check on the index version? (available in the index settings)

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 2]
```diff
@@ -271,6 +271,9 @@ public boolean requiresAnalysisSettings() {
             Settings currentSettings = entry.getValue();
             String typeName = currentSettings.get("type");
             if (analyzer) {
+                if(defaultInstance.containsKey(name)) {
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: I think we should do this for all analysis components, so I would move this test outside of the `if (analyzer)` block. Can you also add one space between your if and the opening parenthesis so that indentation is consistent with the rest of the code base?

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java [hunk 3]
```diff
@@ -271,6 +271,9 @@ public boolean requiresAnalysisSettings() {
...
+                    throw new IllegalArgumentException("do not allow custom analyzers to have the same names [" + name + "] as built-in analyzers");
```
Comment (jpountz) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: Maybe rephrase to something like this: "Custom analysis component [" + name + "] may not reuse the name of a built-in component"

**Review by jpountz — CHANGES_REQUESTED** [review]
Thans @makeyang, I left some comments.

File: core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java (hunk 1 above)
Comment (makeyang) [inline core/src/main/java/org/elasticsearch/index/analysis/AnalysisRegistry.java]: @jpountz only apply this for version above 6.0.0-alpha1?

**Comment by makeyang** [comment]
@jpountz  can you help to review this please?

**Comment by makeyang** [comment]
@jpountz  can you help to reivew this? thanks.

**Comment by rjernst** [comment]
I think this may have already been fixed by the work to move analysis components into a module. @nik9000 can you confirm?

**Comment by nik9000** [comment]
> I think this may have already been fixed by the work to move analysis components into a module. @nik9000 can you confirm?

I can't out of memory. It is certainly possible I fixed it without meaning to.

**Comment by rjernst** [comment]
@makeyang If you are still interested in getting this PR in, can you bring it up to date with master? Can you also add tests for each type of analysis component, since the check now affects them all?

**Comment by makeyang** [comment]
@rjernst got something else to busy with. so forget about me.

**Comment by dakrone** [comment]
Closing as there has been no feedback, feel free to comment if this was closed erroneously.
